### ISSUE TITLE
Twisted web client doesn't include port in the host header

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3857]
|<img alt="Vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Vultaire reported|
|-|-|
|Trac ID|trac#3857|
|Type|defect|
|Created|2009-05-28 05:15:38Z|
|Branch|https://github.com/twisted/twisted/tree/getpage-host-port-3857|

The host header must contain the port number if different from 80 or 443.

This one bit me when I relocated a WordPress blog from port 80 to port 8080 on a server, and was trying to use getPage to grab a page from the blog.  (getPage indirectly calls HttpClient.)

The HttpClient protocol does not seem to properly specify the "host" header in HTTP requests, and because of doing so, is ending up in an infinite redirect loop.

Specifically, given the url "http://domu-XX-XX-XX-XX-XX-XX.compute-1.internal:8080/wordpress/", here's what happens:
```
sendCommand: GET /wordpress/ HTTP/1.0
sendHeader: Host: domu-XX-XX-XX-XX-XX-XX.compute-1.internal
sendHeader: User-Agent: Twisted PageGetter
endHeaders

lineReceived: HTTP/1.0 301 Moved Permanently
lineReceived: Date: Thu, 28 May 2009 04:47:19 GMT
lineReceived: Server: Apache/2.2.9 (Ubuntu) PHP/5.2.6-2ubuntu4.2 with Suhosin-Patch
lineReceived: X-Powered-By: PHP/5.2.6-2ubuntu4.2
lineReceived: X-Pingback: http://domU-XX-XX-XX-XX-XX-XX.compute-1.internal:8080/wordpress/xmlrpc.php
lineReceived: Location: http://domu-XX-XX-XX-XX-XX-XX.compute-1.internal:8080/wordpress/
lineReceived: Vary: Accept-Encoding
lineReceived: Content-Length: 0
lineReceived: Connection: close
lineReceived: Content-Type: text/html; charset=UTF-8
lineReceived: 

[...loops, raises exception, exits.]
```

Connecting via telnet to port 8080 and doing the same commands by hand generates the same response, of course.

However, if I tack ":8080" to the host header:

```
sendCommand: GET /wordpress/ HTTP/1.0
sendHeader: Host: domu-XX-XX-XX-XX-XX-XX.compute-1.internal:8080
sendHeader: User-Agent: Twisted PageGetter
endHeaders

lineReceived: HTTP/1.0 200 OK
lineReceived: Date: Thu, 28 May 2009 05:08:32 GMT
lineReceived: Server: Apache/2.2.9 (Ubuntu) PHP/5.2.6-2ubuntu4.2 with Suhosin-Patch
lineReceived: X-Powered-By: PHP/5.2.6-2ubuntu4.2
lineReceived: X-Pingback: http://domU-XX-XX-XX-XX-XX-XX.compute-1.internal:8080/wordpress/xmlrpc.php
lineReceived: Vary: Accept-Encoding
lineReceived: Connection: close
lineReceived: Content-Type: text/html; charset=UTF-8
lineReceived:

[output follows]
```

[RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23) does specify that the port should be specified on this header, else port 80 will be implied.

I know not all cases are affected; Apache seems to be pretty forgiving and I can retrieve static pages fine.  But when I go for anything in WordPress, then I run into problems.

Attachments:

* [0001-Add-the-port-to-the-Host-header-if-it-is-not-the-def.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/905/905ab9811bede5cf6f3ce6b42495a8c42870f379/6fe30cc59bd8fb6bfb38026203dd7e6fb324711f.patch) (6376 bytes) - added by armooo on 2011-02-27 20:00:28Z - HTTPPageGetter.connectionMade now sends the port in the host header if required.

<details><summary>Searchable metadata</summary>

```
trac-id__3857 3857
type__defect defect
reporter__Vultaire Vultaire
priority__normal normal
milestone__ 
branch__branches_getpage_host_port_3857 branches/getpage-host-port-3857
branch_author__armooo armooo
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1243487738000000 1243487738000000
changetime__1299283552000000 1299283552000000
version__None None
owner__exarkun exarkun
cc__ivank cc__armooo
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Vultaire commented|
|-|-|

I should clarify myself: The problem appears to be in the implementation of HttpPageGetter's connectionMade function.

Specifically, [line 39 of client.py](http://twistedmatrix.com/trac/browser/tags/releases/twisted-8.2.0/twisted/web/client.py#L39):
self.sendHeader('Host', self.factory.headers.get("host", self.factory.host))

If host is not explicitly set, then it defaults to the host stored in the factory.  However, if the port is not 80, it should also be attached in order to be RFC compliant.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank commented|
|-|-|

If http:// and port 80, don't append `:80`. If https:// and port 443, don't append `:443`. Else, append `:port`

(I'm pretty sure this is what everyone else does.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Replying to [ivank](#note_3):
> If http:// and port 80, don't append `:80`. If https:// and port 443, don't append `:443`. Else, append `:port`
>
> (I'm pretty sure this is what everyone else does.)

Um?  I guess you're describing a solution to whatever the problem here is.  I don't really understand what the problem is though.  Is it related to [#3845](https://github.com/twisted/twisted/issues/3845) (another ticket I don't really understand)?

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

The problem is that twisted.web.client never appends a `:port` to the `Host` header, even when it is required to. It is required when the port is not the default port for the protocol.

Yes, this looks like a duplicate of [#3845](https://github.com/twisted/twisted/issues/3845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

Okay, duplicate.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Interestingly, the bug is not present in the Agent code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [30880]) Branching to 'getpage-host-port-3857'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [30881]) Apply 0001-Add-the-port-to-the-Host-header-if-it-is-not-the-def.patch

refs [#3857](https://github.com/twisted/twisted/issues/3857)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Thanks for the patch!  I want to make a few minor tweaks to it:

  1. Methods should be separated from each other by two blank lines
  1. Classes should be separated from each other by three blank lines
  1. There should be one more space in `foo %(bar)` - like `foo % (bar)`
  1. Rather than using `FakeTransport`, the new tests should use `StringTransport`.  In fact, we should get rid of `FakeTransport` and change the one test method that uses it to use `StringTransport` instead (separate ticket).  Better to have one good test double than lots of not very good ones.
  1. Instead of setting the `transport` attribute on a protocol and then calling its `connectionMade` method, you should call its `makeConnection` method with the transport and leave the rest up to it.  This is what happens when a real connection is made.
  1. Indentation of subsequent lines of a multiline function call should line up with the opening `(` of the call, or no args should be on the call line and all lines should be indented 4 spaces.

Nothing significant in this feedback, so I'll check in these changes and then merge the branch.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30882]) Tweaks, mostly to whitespace, but also add a docstring and use `StringTransport` and `makeConnection`

refs [#3857](https://github.com/twisted/twisted/issues/3857)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [30883]) Merge getpage-host-port-3857

Author: armooo
Reviewer: exarkun
Fixes: [#3857](https://github.com/twisted/twisted/issues/3857)

Make `HTTPPageGetter` send _Host_ headers which include the port number
of the server if it is not the default port.
